src/modules/mahjong/HandValidator/HandValidator.ts

```diff
@@ -0,0 +1,246 @@
+
+  /**
+   * Creates an object mapping of the tiles to determine the number of each tile in the hand
+   * @param tiles An array of tile objects
+   * @returns object containing the number of each tile in the hand
+   */
+  public static createTileMapping(tiles: Tile[]): { [index: string]: number } {
+    const mapping: { [index: string]: number } = {};
+    tiles.forEach((t) => {
+      if (Object.prototype.hasOwnProperty.call(mapping, t.toString())) mapping[t.toString()] += 1;
+      else mapping[t.toString()] = 1;
+    });
+
+    return mapping;
+  }
+
+  public static determineAllPossiblePairs(tiles: Tile[]): ValidPair[] {
+    const tileMapping = this.createTileMapping(tiles);
+
+    /**
+     * Valid Pair Schema:
+     * {
+     *  pair: '1_CHARACTER',
+     *  remainingTiles: {},
+     *  numTiles: number
+     * }
+     */
+    const validPairs: ValidPair[] = [];
+    const mappingKeys = Object.keys(tileMapping);
+    const numTiles = Object.values(tileMapping).reduce((pv, cv) => pv + cv);
+
+    mappingKeys.forEach((key) => {
+      if (tileMapping[key] >= 2) {
+        const mappingCopy = { ...tileMapping };
+
+        mappingCopy[key] -= 2;
+
+        validPairs.push({
+          pair: key,
+          remainingTiles: mappingCopy,
+          numTiles,
+        });
+      }
+    });
+
+    return validPairs;
+  }
+
+  /**
+   * Determines all possible valid hands and invalid hands
+   * @param validPairs an array of ValidPairs
+   */
+  public static validiateValidHandStructure(validPairs: ValidPair[]): HandStructureResults {
+    const results: HandStructureResults = {
+      valid: [],
+      invalid: [],
+    };
+
+    validPairs.forEach((vp) => {
+      const { remainingTiles, numTiles } = vp;
+      const copyRemainingTiles = { ...remainingTiles }; // modify copy to preserve original object
+      const melds: Meld[] = [];
+      let passed = true;
+
+      if (numTiles > this.MIN_HAND_SIZE) {
+        // There is a 4 of a kind somewhere in the hand and has to be used as a four of a kind
+        Object.keys(copyRemainingTiles).forEach((key) => {
+          if (copyRemainingTiles[key] === this.KAN_SIZE) {
+            const meld: Meld = {
+              tiles: [key, key, key, key],
+              type: MeldTypes.QUAD,
+            };
+            copyRemainingTiles[key] -= this.KAN_SIZE;
+            melds.push(meld);
+          }
+        });
+      }
+
+      // Start trying to create all consecutive melds
+      // Cannot start with triplets as there is a chance we use the triplet for a consecutive
+      Object.keys(copyRemainingTiles).forEach((key) => {
+        const isSimpleTile: boolean = Object.values(SimpleTileTypes).includes(<SimpleTileTypes>TileMapper[key].type);
+
+        // n = number that can be used to create a consecutive
+        let n = copyRemainingTiles[key] < this.TRIPLET_SIZE || copyRemainingTiles[key] === this.KAN_SIZE;
+        n = copyRemainingTiles[key] > 0 && n;
```

Comment:
What's the functions of these 4 booleans to n

Reply:
This is used to create a consecutive. What it'll do is it'll start at one tile and check how many tiles of that type there are. If there are less than 3 or 4, then it's eligible to create a consecutive and we'll add that to the meld. If it's 3, then we shouldn't use that because it needs to be used as a triplet